R/rma-tidiers.R

```diff
@@ -0,0 +1,222 @@
+                     include_studies = TRUE, measure = "GEN", ...) {
+  
+  estimates <- metafor::escalc(yi = x$yi.f, vi = x$vi.f, measure = measure) %>%
+    summary(level = conf.level * 100) %>% 
+    as.data.frame(stringsAsFactors = FALSE)
+  
+  estimates <- cbind(x$slab, "study", estimates[, c("yi", "sei", "zi")], NA,
+                     estimates[, c("ci.lb", "ci.ub")], stringsAsFactors = FALSE)
+  names(estimates) <- c("study", "type", "estimate", "std.error", "statistic",
+                        "p.value", "conf.low", "conf.high")
+  estimates <- tibble::as_tibble(estimates)
+  
+  betas <- x$beta
+  if (!is.null(nrow(betas)) && nrow(betas) > 1) {
+    # get estimate type and fix spelling
+    study <- rownames(betas)
+    swap <- grepl("intrcpt", study)
+    study[swap] <- "intercept"
+    betas <- as.double(betas)
+  } else {
+    study <- "overall"
+    betas <- betas[1]
+  }
+  
+  results <- tibble::tibble(study = study, type = "summary",
+                            estimate = betas, std.error = x$se,
+                            statistic = x$zval, p.value = x$pval,
+                            conf.low = x$ci.lb, conf.high = x$ci.ub)
+  .data <- if (include_studies) rbind(estimates, results) else results
+  
+  if (exponentiate) {
+    .data$estimate <- exp(.data$estimate)
+    .data$conf.low <- exp(.data$conf.low)
+    .data$conf.high <- exp(.data$conf.high)
+  }
+  
+  if (!conf.int) {
+    .data <- .data[-which(names(.data) %in% c("conf.low", "conf.high"))]
+  }
+  
+  attributes(.data$study) <- NULL
+  
+  tibble::remove_rownames(.data)
+}
+
+#' Construct a one-row summary of meta-analysis model fit statistics.
+#'
+#' `glance()` computes a one-row summary of  meta-analysis objects,
+#'  including estimates of heterogenity and model fit.
+#'
+#' @param x An `rma` created by the `metafor` package.
+#' @param ... Additional arguments
+#'
+#' @return a `tibble`
+#' @export
+#'
+#' @examples
+#'
+#' library(metafor)
+#' 
+#' df <-
+#'   escalc(
+#'     measure = "RR",
+#'     ai = tpos,
+#'     bi = tneg,
+#'     ci = cpos,
+#'     di = cneg,
+#'     data = dat.bcg
+#'   )
+#' 
+#' meta_analysis <- rma(yi, vi, data = df, method = "EB")
+#'
+#' glance(meta_analysis)
+#'
+glance.rma <- function(x, ...) {
+  fit_stats <- metafor::fitstats(x)
+  fit_stats <- fit_stats %>%
+    t() %>%
+    as.data.frame()
+  names(fit_stats) <-
+    stringr::str_replace(names(fit_stats), "\\:", "")
+  
+  list(
+    nobs = x$k,
+    measure = x$measure,
+    method = x$method,
+    i.squared = x$I2,
+    h.squared = x$H2,
+    tau.squared = x$tau2,
+    tau.squared.se = x$se.tau2,
+    qe = x$QE,
+    p.value.qe = x$QEp,
+    qm = x$QM,
+    p.value.qm = x$QMp,
+    fit_stats
+  ) %>%
+    # get rid of null values
+    purrr::discard(is.null) %>%
+    # don't include multivariate model stats
+    purrr::discard(~length(.x) >= 2) %>%
+    # change to tibble with correct column and row names
+    as.data.frame() %>%
+    tibble::as_tibble() %>%
```

Comment:
I don't think `tibble::remove_rownames()` is necessary if you use `tibble::as_tibble()`?

Reply:
You're right! These are leftover from when I first wrote these, which was prior to broom using tibble